jormungandr/src/explorer.rs

```diff
@@ -0,0 +1,86 @@
+use super::blockchain::Blockchain;
+use crate::blockcfg::Block;
+use crate::blockcfg::ChainLength;
+use crate::intercom::ExplorerMsg;
+use crate::utils::task::{Input, ThreadServiceInfo};
+use chain_impl_mockchain::multiverse::{GCRoot, Multiverse};
+use futures::Future;
+use futures::IntoFuture;
+
+use chain_core::property::Block as _;
+
+pub struct Process {
+    multiverse: Multiverse<State>,
+}
+
+struct State {
+    block: Block,
+}
+
+impl State {
+    pub fn new(block: Block) -> Self {
+        Self { block }
+    }
+}
+
+impl Process {
+    pub fn new() -> Self {
+        Self {
+            multiverse: Multiverse::<State>::new(),
+        }
+    }
+
+    pub fn handle_input(
+        &mut self,
+        info: &ThreadServiceInfo,
+        blockchain: &Blockchain,
```

Comment:
Since you have access to the `Blockchain` you also have access to the `RefCache` and the `Ref`. So something interesting is that you don't need to store the block in the explorer `State`. You can simply hold a `Ref` and query the `Storage` for the content of the block.

I.e., something you can do instead is to have the following:

```rust
struct Process {
    multiverse: Multiverse<Ref>, // storage by block hash

    chain_length_to_hash: HashMap<ChainLength, Vec<Ref>>,
}
```

and to have the `Ref` passed on by the `ExplorerMsg`'s `NewBlock`.